File: OPF_Python/opf_sdp_cvx.py
#%%
import cvxpy as cp
import numpy as np
from math import radians

#%%
# from src.struct_network import Bus, Line, Generator
from src.readnetwork import readnetwork
from src.ybus import ybus
from src.sdp_ymat import sdp_ymat


testsystem = "case_ieee30"

#%%
buses, lines, generators, datamat = readnetwork(testsystem)

nline = len(lines)
nbus = len(buses)
ngen = len(generators)
lineset = np.arange( 0, nline )
busset = np.arange( 0, nbus )
genset = np.arange(0, ngen )
baseMVA = datamat["baseMVA"][0, 0][0, 0]

B_g = []
for g in genset:
    B_g.append( generators[g].location )

B_gn = []
for i in busset:
    B_gn.append( [] )

for g in genset:
    B_gn[ generators[g].location ].append( int(g) )


#%%
Ybus, yff, yft, ytf, ytt = ybus( buses, lines )
Yk, Yk_, Mk, Ylineft, Ylinetf, Y_lineft, Y_linetf, YL, YL_ = sdp_ymat(
    lines, Ybus)


gcostdata = datamat["gencost"][0, 0]

if np.size(gcostdata, 2-1) == 7:
    c_k2 = gcostdata[:, 5-1] * baseMVA ** 2
    c_k1 = gcostdata[:, 6-1] * baseMVA
    c_k0 = gcostdata[:, 7-1].T
elif np.size(gcostdata, 2-1) == 6:
    c_k2 = np.zeros(np.size(gcostdata, 1-1), 1)
    c_k1 = gcostdata[:, 5-1] * baseMVA
    c_k0 = gcostdata[:, 6-1].T

#%%
# p5 right column: a small resistance (1e-5) is added to each transformer ...
# ∴ connected graph
# for l in lineset:
#     if lines[l].r == 0:
#         lines[l].r == 1e-5;
#%% cvx start

# W = cp.Variable((2*len(busset), 2*len(busset)), symmetric=True)
W = cp.Variable((2*len(busset), 2*len(busset)), PSD=True)
alpha_k = cp.Variable(len(genset))

constraints = []
for b in busset:
    if buses[b].btype == 3:
        constraints += [W[:,nbus+b] == 0]
        constraints += [W[nbus+b,:] == 0]
        slack_bus = b;
        print("slack bus:", slack_bus)

for b in busset:
    # Equations (4a)
    constraints += [sum(generators[g].Pmin for g in B_gn[b]) - buses[b].Pd <= cp.trace(Yk(b) @ W)]
    constraints += [sum(generators[g].Pmax for g in B_gn[b]) - buses[b].Pd >= cp.trace(Yk(b) @ W)]
    # Equations (4b)
    constraints += [sum(generators[g].Qmin for g in B_gn[b]) - buses[b].Qd <= cp.trace(Yk_(b) @ W)]
    constraints += [sum(generators[g].Qmax for g in B_gn[b]) - buses[b].Qd >= cp.trace(Yk_(b) @ W)]
    # Equations (4c)
    constraints += [(buses[b].Vmin)**2 <= cp.trace(Mk(b) @ W)]
    constraints += [(buses[b].Vmax)**2 >= cp.trace(Mk(b) @ W)]


for l in lineset:
    if lines[l].u != 0:
        # Equations (4e)
        constraints += [- lines[l].u <= cp.trace(Ylineft(l) @ W)]
        constraints += [+ lines[l].u >= cp.trace(Ylineft(l) @ W)]
        # Equations (5): 3x3 matrix
        constraints += [
            cp.bmat([
                [-(lines[l].u)**2, cp.trace(Ylineft(l) @ W), cp.trace(Y_lineft(l) @ W)],
                [cp.trace(Ylineft(l) @ W), -1, 0],
                [cp.trace(Y_lineft(l) @ W), 0, -1]
            ]) << 0]

for g in genset:
    # Equations (6): 2x2 matrix
    constraints += [
        cp.bmat([
            [
                c_k1[g] * cp.trace(Yk(B_g[g]) @ W) - alpha_k[g] + c_k0[g] + c_k1[g] * buses[B_g[g]].Pd,
                cp.sqrt(c_k2[g]) * cp.trace(Yk(B_g[g]) @ W) + cp.sqrt(c_k2[g]) * (buses[B_g[g]].Pd)
            ],
            [
                cp.sqrt(c_k2[g]) * cp.trace(Yk(B_g[g]) @ W) + cp.sqrt(c_k2[g]) * (buses[B_g[g]].Pd), 
                -1
            ]
        ]) << 0]
    

#%%
prob = cp.Problem(cp.Minimize(sum(alpha_k[g] for g in genset)), constraints)

#%%
# print(cp.installed_solvers())
# prob.solve(solver=cp.GUROBI, verbose=True)
prob.solve( solver=cp.MOSEK , verbose = True )
# prob.solve( verbose = True )
print("Termination status =", prob.status)
print("The optimal value is", prob.value)
print("test system =", testsystem)
print("solve time =", prob.solver_stats.solve_time)


#%%
v = [np.sqrt(np.trace(Mk(b) @ W.value)) for b in busset]
fp = [np.trace(Ylineft(l) @ W.value) for l in lineset]
fq = [np.trace(Y_lineft(l) @ W.value) for l in lineset]
pg = [np.trace(Yk(B_g[g]) * W.value) + buses[B_g[g]].Pd for g in genset]
qg = [np.trace(Yk_(B_g[g]) * W.value) + buses[B_g[g]].Qd for g in genset]


# for variable in prob.variables():
#     print("Variable %s: value %s" % (variable.name(), variable.value))

File: OPF_Matlab/opf_sdp_cvx.m
clear all
testsystem = "case_ieee30";
load("data/network/mat/" + testsystem + ".mat"); % load function from MATPOWER




[buses, lines, generators] = readnetwork(mpc);


nline = length(lines);
nbus = length(buses);
ngen = length(generators);

lineset = 1:nline;
busset = 1:nbus;
genset = 1:ngen;


B_g = [];
for g = genset
    B_g = [ B_g; generators(g).location ];
end

%%
B_gn_cell = cell( nbus, 1 );
for g = genset
    B_gn_cell{generators(g).location} = [B_gn_cell{generators(g).location}; g];
end
B_gn = containers.Map( busset, B_gn_cell);


[ Ybus, yff, yft, ytf, ytt ] = ybus( buses, lines );


% Add resistance to transformers/lines -> connected resistive graph
% for l = lineset
%     if lines(l).r == 0
%         lines(l).r = 0.0001;
%     end
% end

[Yk,Yk_,Mk,Ylineft,Ylinetf,Y_lineft,Y_linetf,YL,YL_] = sdp_ymat( lines, Ybus );



cvx_solver Mosek
cvx_begin SDP
variable W( 2*nbus, 2*nbus ) semidefinite
variable alpha_k( ngen, 1 )
minimize sum( alpha_k, 1 )
% slack bus
for b = busset
    if buses(b).btype == 3
        W( :, nbus+b ) == 0;
        W( nbus+b, : ) == 0;
        slack_bus = b;
        display("slack bus: " + slack_bus)
    end
end

for b = busset
    sum([generators(B_gn(b)).Pmin]) - buses(b).Pd <= trace( Yk(b) * W ) <= sum([generators(B_gn(b)).Pmax]) - buses(b).Pd
    sum([generators(B_gn(b)).Qmin]) - buses(b).Qd <= trace( Yk_(b) * W ) <= sum([generators(B_gn(b)).Qmax]) - buses(b).Qd
end

% Equation (4c)
for b = busset
    ( buses(b).Vmin )^2 <= trace( Mk(b) * W ) <= ( buses(b).Vmax )^2
end
% Equation (4e)
for l = lineset
    if lines(l).u ~= 0
        - lines(l).u <= trace( Ylineft(l) * W ) <= lines(l).u
    end
end
% Equations (5)
for l = lineset
    if lines(l).u ~= 0
        [...
            -( lines(l).u )^2, trace( Ylineft(l) * W ), trace( Y_lineft(l) * W );...
            trace( Ylineft(l) * W ), -1, 0;...
            trace( Y_lineft(l) * W ), 0, -1] <= 0
    end
end
% Equation (6)
for g = genset
    [...
        generators(g).cost(2) * ( trace( Yk(B_g(g)) * W ) * mpc.baseMVA ) - alpha_k( g, 1 ) + generators(g).cost(3) + generators(g).cost(2) * ( buses(B_g(g)).Pd * mpc.baseMVA ), ...
        sqrt( generators(g).cost(1) ) * ( trace( Yk(B_g(g)) * W ) * mpc.baseMVA ) + sqrt( generators(g).cost(1) ) * ( buses(B_g(g)).Pd * mpc.baseMVA ); ...
        sqrt( generators(g).cost(1) ) * ( trace( Yk(B_g(g)) * W ) * mpc.baseMVA ) + sqrt( generators(g).cost(1) ) * ( buses(B_g(g)).Pd * mpc.baseMVA ), -1] <= 0
end
cvx_end

cvx_objvalue = sum(alpha_k,1);

%%
